modules/metastorage/src/integrationTest/java/org/apache/ignite/internal/metastorage/impl/ItIdempotentCommandCacheTest.java

```diff
@@ -334,8 +325,45 @@ public void testIdempotentInvokeAfterLeaderChange() {
         assertThat(futAfterLeaderChange, willCompleteSuccessfully());
         assertTrue(futAfterLeaderChange.join());
 
-        assertTrue(currentLeader.checkValueInStorage(testKey.bytes(), testValue));
-        assertTrue(newLeader.checkValueInStorage(testKey.bytes(), testValue));
+        assertTrue(currentLeader.checkValueInStorage(TEST_KEY.bytes(), TEST_VALUE));
+        assertTrue(newLeader.checkValueInStorage(TEST_KEY.bytes(), TEST_VALUE));
+    }
+
+    @ParameterizedTest
+    @MethodSource("idempotentCommandProvider")
+    public void testIdempotentCacheRestoreFromSnapshot(IdempotentCommand idempotentCommand, TestInfo testInfo) throws Exception {
+        RaftGroupService raftClient = raftClient();
+
+        // Initial idempotent command run.
+        CompletableFuture<Object> commandProcessingResultFuture = raftClient.run(idempotentCommand);
+        assertThat(commandProcessingResultFuture, willCompleteSuccessfully());
+        Object commandProcessingResult = commandProcessingResultFuture.get();
+        if (commandProcessingResult instanceof Boolean) {
+            assertTrue((Boolean) commandProcessingResult);
+        } else {
+            assertTrue(((StatementResult) commandProcessingResult).getAsBoolean());
+        }
+
+        // Do the snapshot.
+        nodes.forEach(n -> raftClient().snapshot(new Peer(n.clusterService.nodeName())));
+
+        // Restart nodes in order to trigger idempotent volatile cache initialization from snapshot.
+        for (Node node : nodes) {
+            node.stop();
+        }
+
+        // Restart cluster.
+        startCluster(testInfo);
+
+        // Run same idempotent command one more time and check that condition wasn't re-evaluated, but was retrieved from the cache instead.
+        CompletableFuture<Object> commandProcessingResultFuture2 = raftClient().run(idempotentCommand);
+        assertThat(commandProcessingResultFuture2, willCompleteSuccessfully());
+        Object commandProcessingResult2 = commandProcessingResultFuture2.get();
+        if (commandProcessingResult2 instanceof Boolean) {
+            assertTrue((Boolean) commandProcessingResult2);
+        } else {
+            assertTrue(((StatementResult) commandProcessingResult2).getAsBoolean());
+        }
```

Comment:
What about checking the actual values in the storage?

Reply:
Added.